Implement output transform

tdaniel22 opened the pull request:
This commit adds two command line switches. -r rotates the output 90
degrees clockwise and can be specified multiple times. -f flips the
output.

Fixes #55

Hjdskes commented on cage.c:
```diff
@@ -101,14 +103,23 @@ parse_args(struct cg_server *server, int argc, char *argv[])
+				server->output_rotation = WL_OUTPUT_TRANSFORM_NORMAL;
+			}
+			break;
+		case 'f':
```
Honest question: what is the use case for flipping the output?

tdaniel22 replied:
I honestly don't know either .. maybe some weird devices need this? Given that it's part of the transform enum I thought I would make a flag for it. I can remove it if you don't feel like it deserves a dedicated flag.

tdaniel22 replied:
Also, is passing the rotation flag multiple times OK with you? I wanted to avoid doing more parsing and went for the lazy way, but there might be a more elegant way.

Hjdskes replied:
Please remove the flag if there is no good use case for it. It'll be easy to add back if the feature is ever requested.

Passing `-r` multiple times is OK with me. I think it's rather elegant in a way :)

Hjdskes commented on output.c:
```diff
@@ -211,10 +211,7 @@ handle_output_damage_frame(struct wl_listener *listener, void *data)
-	wlr_output_transformed_resolution(output->wlr_output, &output_width, &output_height);
-
-	wlr_renderer_begin(renderer, output_width, output_height);
+	wlr_renderer_begin(renderer, output->wlr_output->width, output->wlr_output->height);
```
Why did you access the output width and height directly instead of through `wlr_output_transformed_resolution`?

tdaniel22 replied:
Apparently `wlr_renderer_begin` expects the real resolution. Rootston does it like this (https://github.com/swaywm/wlroots/blob/master/rootston/render.c#L240). Tinywl does not support output rotation so passing the transformed resolution did not make a difference, but I believe this was a bug.

Hjdskes replied:
Indeed, thanks for catching this.

Hjdskes commented on server.h:
```diff
@@ -39,6 +39,8 @@ struct cg_server {
 #endif
 
 	bool xdg_decoration;
+	int output_rotation;
```
Please use the proper `enum wl_output_transform` type here. And perhaps rename `output_rotation` to `output_transform` instead.

Hjdskes commented:
Thanks and sorry for not getting back to you earlier!

tdaniel22 commented:
This is ready for a second review.

flokli commented:
Can we get this to also rotate inputs? I use this on a touchscreen, and the touch input doesn't rotate currently.

tdaniel22 commented:
Well that's weird, at the time I implemented this touch was rotated correctly. But now that I've updated my buildroot system I'm having this problem as well. The touch input handling code in cage and wlroots hasn't changed since then AFAICT, which is where I would have expected this to be done. Mouse input is still correct though.

tdaniel22 commented:
See #126
